[ABI] Use mangled names for base protocol witnesses.

DougGregor opened the pull request:
Emit resilient witnesses for base protocols as mangled names, using the same
information as for associated conformances. Teach the runtime to eagerly
realize these witnesses during instantiation so clients of inherited witness
tables are not affected. There are a few benefits to these changes:

* The runtime now copies the arguments corresponding to conditional 
requirements into the private area of the witness table, so we don't have
to hardcode that operation into the compiler-generated instantiation
function. (Minor code size improvement)
* Resilient witnesses can be more compact.
* We can get the benefits of mangling conformances (vs. open-coding them)
once we've implemented that, e.g., smaller code size, ability 
* Base protocol witness tables are handled the same way as associated conformance
witness tables, making both easier to work with. This also could allow, e.g., a
default implementation of a base protocol witness table should we need that in the
future.
* Instantiation functions are only needed for non-resilient conformances with
non-constant witness tables.
* Resilient witness tables no longer need a witness table "pattern", since there
wasn't really much information in it anyway.

Implements rdar://problem/46282080.

DougGregor committed:
[Runtime] Eagerly realize mangled base protocol witnesses.

If the witness for a base protocol is a mangled name, eagerly turn that name
into a witness table via swift_getAssociatedConformanceWitness(). This
allows the compiler to emit base protocol witnesses as mangled names
(as it does for associated conformances).

DougGregor committed:
[Mangling] Generalize the mangling of associated witness table accessors.

Use a general ‘type’ production for the conforming type of an associated
witness table accessor mangling, so that we can mangle base protocol
witness table accessors. These entities are always internal symbols, so the
mangling itself doesn’t affect the ABI.

DougGregor committed:
[IRGen] Emit base protocol witnesses as mangled names.

Rather than setting the witness tables for base protocols in a witness
table’s instantiation function, use mangled names as we do for
associated conformances. The runtime will eagerly realize them so clients
of witness tables are not affected.

For now, only do this for unconditional conformances: conditional
conformances won’t work yet because the instantiation function currently
copies the instantiation arguments (which satisfy the conditional
requirements) in to the private area. The runtime will need to take over
this responsibility for conditional conformances to work.

Part of rdar://problem/46282080.

DougGregor commented:
@swift-ci please benchmark

DougGregor committed:
[Runtime] Have the runtime copy instantiation arguments into private data.

Rather than having the witness table instantiation function copy the 
instantiation arguments that corresponding to conditional requirements
into the private area, have the runtime do it. We’re going to depend on
these lining up anyway (and the witness table can of course have *more*
private slots that it manages some other way), so centralize the code.

More of rdar://problem/46282080.

DougGregor committed:
[IRGen] Enabled mangled inherited conformances for conditional conformances.

Now that the runtime fills in the private slots with conditional requirement
information, we can always use mangled names for inherited conformances.

Fixes rdar://problem/46282080.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please benchmark

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please benchmark

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[ABI] Emit associated conformance descriptors for inherited protocols.

Start emitting associated conformance requirement descriptors for
inherited protocols, so we have a symbol to reference from resilient
witness tables and mangled names in the future.

DougGregor committed:
[ABI] Put mangled inherited protocol witnesses into the resilient witness table

Emit mangled inherited protocol witnesses into the resilient witness table,
and realize them when we instantiate the resilient witness table. Don’t
put mangled inherited protocol witnesses into non-resilient witness tables:
there is no efficient way to make sure they are realized, so keep the
previous instantiation-function approach.

Implements the rest of rdar://problem/46282080.

DougGregor committed:
[ABI] Optimize out the witness table pattern for resilient conformances.

Resilient conformances now put all witnesses into the resilient witness
table, so optimize away the witness table pattern for such cases. Teach
the runtime to fill in the protocol conformance descriptor, because that’s
the only bit of real information that would be in the pattern.

This is a minor optimization enabled by rdar://problem/46282080.

rjmccall commented on docs/ABI/Mangling.rst:
```diff
@@ -146,7 +146,8 @@ Globals
   global ::= protocol-conformance identifier 'Wt' // associated type metadata accessor (HISTORICAL)
-  global ::= protocol-conformance assoc-type-list nominal-type 'WT' // associated type witness table accessor
+  global ::= protocol-conformance assoc-type-list protocol 'WT' // associated type witness table accessor
+  global ::= protocol-conformance type protocol 'WT' // inherited protocol witness table accessor
```
Is there no ambiguity here?

Also, why do we need to define an accessor for this?  Has the resilient code generation for accessing a base protocol changed from just a load?

DougGregor replied:
The accessor is there so we can call it to fill in the witness table entry, and is referenced via the mangled name (the same as we do for associated conformances). This moves the code from the instantiation function into the accessor.

The resilient code generation for accessing a base protocol has not changed, because the runtime forces the witness table entries for base protocols to be realized during instantiation.

DougGregor replied:
It's an internal-linkage function that returns the right witness table.

rjmccall commented on docs/ABI/Mangling.rst:
```diff
@@ -199,6 +200,7 @@ types where the metadata itself has unknown layout.)
   global ::= assoc-type-name 'Tl'        // associated type descriptor
   global ::= assoc-type-name 'TM'        // default associated type witness accessor (HISTORICAL)
   global ::= type assoc-type-list protocol 'Tn' // associated conformance descriptor
+  global ::= type type protocol 'Tn' // inherited protocol witness descriptor
```
Same question about ambiguity.  Also, I think a better name for this would be "base conformance descriptor", as saying "protocol witness" invites confusion with the more general concept.

DougGregor replied:
I think it's unambiguous, but I agree that it's safer to simply burn another two mangling characters here. Will follow up once testing here completes.

DougGregor committed:
[Test] Update FileCheck test for ELF/COFF

rjmccall commented on lib/Demangling/Demangler.cpp:
```diff
@@ -1875,6 +1875,20 @@ NodePointer Demangler::popAssocTypePath() {
+    pushNode(Type);
+  }
+
+  return popAssocTypePath();
```
Yeah, I mean, I don't know if reducing the number of mangling operators really justifies the complexity and risk of future conflict here.

DougGregor replied:
ok

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please benchmark

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please benchmark

rjmccall commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -729,6 +729,20 @@ namespace {
         }
 
+        if (entry.isBase()) {
+          // Define the associated conformance descriptor to point to the
```
"Define a base conformance descriptor, which is just an associated conformance descriptor which etc."

DougGregor committed:
[Mangling] Separate out base conformance descriptors.

Separate the mangling of base conformance descriptors from that of
associated conformance descriptors, and simplify it.

DougGregor committed:
[Mangling] Separate mangling for associated/base conformance accessors.

It’s simpler to use separate manglings for these two cases.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility
